story/MHDP-5

prnvshrn opened the pull request:
Added functionality to create, edit and delete investigator records

jdamerow commented on browser/forms.py:
```diff
@@ -329,3 +329,9 @@ def save(self, user):
         return new_person
+
+
+class InvestigationForm(forms.ModelForm):
```
this should be called "InvestigatorForm" according to its model class

prnvshrn replied:
Changed the name to "InvestigatorForm" from "InvestigationForm"

jdamerow commented on browser/templates/browser/base.html:
```diff
@@ -98,8 +98,22 @@
+                $(".modal-body #investigator_delete_modal_subject").text(subject);
+
+                });
+
```
this should not be in the base template. the base template should contain code needed by all pages, not just one.

prnvshrn replied:
Removed the query code from base.html and inserted them into the respective html files

jdamerow commented on browser/templates/browser/investigator_delete_modal.html:
```diff
@@ -0,0 +1,27 @@
+        <strong>Year :</strong> <label style="font-weight:normal" id="investigator_delete_modal_year"></label> <br>
+        <strong>Subject :</strong> <label style="font-weight:normal" id="investigator_delete_modal_subject"></label> <br>
+        <strong>Role :</strong> <label style="font-weight:normal" id="investigator_delete_modal_role"></label><br>
+        <input type="text" style="display:none;" id="investigator_delete_modal_id"></input><br>
```
there is no blank before colons

jdamerow replied:
input type=hidden should be used if you don't want the input field to be displayed

prnvshrn replied:
Removed the blank before colons, and also the hidden-styled input element.

jdamerow commented on browser/templates/browser/investigator_delete_modal.html:
```diff
@@ -0,0 +1,27 @@
+      <div class="modal-footer">
+        <a
+        href=""
+        onclick="this.href='investigator/'+document.getElementById('investigator_delete_modal_id').value; deletefunction(this, event); alert('Delete');return true;">
```
use jquery for this

jdamerow replied:
and what is the ```alert('Delete')``` doing there?

prnvshrn replied:
Implemented Jquery to implement the functionality. Removed the alert statement.

jdamerow commented on browser/templates/browser/investigator_delete_modal.html:
```diff
@@ -0,0 +1,27 @@
+        href=""
+        onclick="this.href='investigator/'+document.getElementById('investigator_delete_modal_id').value; deletefunction(this, event); alert('Delete');return true;">
+        <button type="button" class="btn btn-danger">Delete</button>
+        </a>
```
there should be either a ```button``` or an ```a``` but not both

prnvshrn replied:
Removed the anchor tag portion, so we are left only with a button

jdamerow commented on browser/templates/browser/investigator_edit.html:
```diff
@@ -0,0 +1,89 @@
+    </div>
+
+</div>
+{% endblock %}
```
you should be able to use the same template for create and upate since the only difference is that there already exist an instance.

prnvshrn replied:
Changed code so that only one template is used to create and update investigator

jdamerow commented on browser/templates/browser/person.html:
```diff
@@ -143,8 +156,18 @@
+                               aria-hidden="true"
+                               href="{% url "edit-investigator" person.id research.id %}">
+                            </a>
+                            <a class="open-InvestigatorDeleteModal btn btn-xs fa fa-trash"
```
why is ```open-InvestigatorDeleteModal``` a class and not an id?

prnvshrn replied:
Changed open-InvestigatorDeleteModal to id

jdamerow commented on browser/views.py:
```diff
@@ -624,3 +624,71 @@ def attendee_create(request, course_id):
+    form = InvestigationForm()
+
+    if request.method == 'POST':
+        if 'SaveButton_Clicked' in request.POST:
```
why do you ask for ```SaveButton_Clicked``` to be present in the post request?

prnvshrn replied:
Removed SaveButton_Clicked from post request , implemented the functionality through JQuery

jdamerow commented on browser/views.py:
```diff
@@ -624,3 +624,71 @@ def attendee_create(request, course_id):
+                investigator.save()
+                return HttpResponseRedirect(reverse('person', args=(person.id,)))
+
+        if 'CancelButton_Clicked' in request.POST:
```
cancel shouldn't be a post request, just have it link back to the person's page.

prnvshrn replied:
Used JQuery to link cancel button to previous page. Removed cancel from post request.

jdamerow commented on browser/views.py:
```diff
@@ -624,3 +624,71 @@ def attendee_create(request, course_id):
+        'investigator_data': research,
+    }
+
+    return render(request, template, context)
```
create and update can be one function

prnvshrn replied:
Is it alright if I keep create and update as separate ? 
This is because, update uses an extra parameter research-id which is not used in the create logic. And a research id cannot be passed as a parameter while creating a investigator data because it has not yet been created.

jdamerow commented:
next time please reopen **this PR**. don't open a new one.

jdamerow commented on browser/templates/browser/investigator.html:
```diff
@@ -0,0 +1,109 @@
+        {% else %}
+        <input class="btn btn-success" name="SaveButton" form="InvestigatorForm" type="submit" value="Save">
+        {% endif %}
+        <input class="btn btn-danger" name="CancelButton" type="button" id="CancelButton" value="Cancel">
```
this can simply be a link (with the correct css class to make it look like a button) and then use django template url to create the link (no funky javascript substring hack).

prnvshrn replied:
Changed the CancelButton to a link. The css gives it the appearance of a button and the link is handled through Django template

jdamerow commented on browser/templates/browser/investigator_delete_modal.html:
```diff
@@ -0,0 +1,24 @@
+      <div class="modal-body">
+
+        <h5>
+        Are you sure you want to delete following data for <strong>{{person.name}}</strong> : <br><br>
```
why would there be a blank between the question and the colon?

prnvshrn replied:
Removed the blank between the question and colon

jdamerow commented on browser/templates/browser/person.html:
```diff
@@ -1,9 +1,38 @@
+            });
+
+            $(document).on("click", "#investigator_delete_modal_deleteButton", function () {
+                window.location = 'investigator/'+ research_id+'/delete/' ;
```
the researcher id should be coming from the clicked button (not as a global variable)

jdamerow replied:
and it needs to be a post request

prnvshrn replied:
I have made the delete investigator as a post request. However, I have retrieved the research_id field in a hidden input text field. Is this an acceptable approach ?

jdamerow replied:
Depends on how you actually implemented it. Just make a new PR when you're done and I'll look at it.

jdamerow commented on browser/views.py:
```diff
@@ -624,3 +624,65 @@ def attendee_create(request, course_id):
+
+@staff_member_required
+def add_investigator_record(request, person_id):
+    print (request)
```
remove

prnvshrn replied:
Removed the statement

jdamerow commented on browser/views.py:
```diff
@@ -624,3 +624,65 @@ def attendee_create(request, course_id):
+
+@staff_member_required
+def delete_investigator_record(request,person_id,research_id):
+    person = get_object_or_404(Person, pk=person_id)
```
needs to be POST

prnvshrn replied:
I have deleted the above method altogether. The response from the post request initiated in the delete investigator modal is now evaluated in the <b>person function</b> at views.py . 

I have modified the person.html so that only users with staff credentials will have the option to edit and delete investigator visible.

jdamerow commented:
Also, please write real commit messages. Just copying the same over and over again is useless.

jdamerow commented on browser/templates/browser/person.html:
```diff
@@ -1,9 +1,29 @@
 {% extends "browser/base.html" %}
```
why would the permissions for this file change?
```100644 -> 100755```

jdamerow replied:
this is still here?

jdamerow commented on browser/views.py:
```diff
@@ -104,6 +104,12 @@ def person(request, person_id=None):
         # context['persons'] = get_paginator(Person, request, 'last_name', 100)
         template = "browser/person_list.html"
+
+    if request.method == 'POST':
```
this should only be allowed if the user is a staff member

prnvshrn replied:
I have added a code in the html file such that only users with staff privileges can see the delete and edit options for investigations. So, only staff users will be able to click on delete button and make the delete post request. 
Is this fine or there should be user validations in the <b>person</b> method in <b>views.py</b> file ?

jdamerow replied:
templates (which generates pages for the client side) and server api are two different things. just because you don't show the button in the browser, doesn't mean the api method isn't open to anyone (the browser is just making the request, another app could make a request, the user could modify the webpage before submitting, etc.).

jdamerow commented on mbl/urls.py:
```diff
@@ -33,6 +33,9 @@
     url(r'^person/$', views.person, name="person_list"),
+    url(r'^person/(?P<person_id>[0-9]+)/investigator/add/$',views.add_investigator_record,name='add-investigator'),
+    url(r'^person/(?P<person_id>[0-9]+)/investigator/(?P<research_id>[0-9]+)/edit/$',views.edit_investigator_record,name='edit-investigator'),
+    url(r'^person/(?P<person_id>[0-9]+)/investigator/delete/$', views.delete_investigator_record, name='delete-investigator'),
```
same as the edit, it should be ```investigator/[id]/delete```

prnvshrn replied:
This line of code <b>extends "browser/base.html</b> is used to form the structure of the html page. So that is why I have not removed it.

I have added <b>investigator/id/delete</b> in the urls.py file.

jdamerow replied:
my comment for person.html is not about the line of code but the file permission. please read my original comment.

prnvshrn replied:
I have changed the permission from 755 to 644 of <b>person.html</b>

jdamerow commented on browser/views.py:
```diff
@@ -624,3 +625,63 @@ def attendee_create(request, course_id):
+
+@staff_member_required
+def delete_investigator_record(request, person_id):
+    Investigator.objects.filter(id=request.POST.get("investigator_delete_modal_id"), person_id=person_id).delete()
```
do you need the person id for this? isn't the investigator id a primary key?

jdamerow replied:
also, what if someone makes a GET request to this method?

prnvshrn replied:
Yes, person_id is not required for deletion because investigator id is the primary key. I have edited the delete method wherein only the research id or investigator id is required.

I logged out of the system and made this request <b>"GET /person/2303/investigator/37804/delete/ HTTP/1.1"</b>. And, I was not able to delete the investigator record and was redirected to the admin page.

jdamerow replied:
the problem here is not being logged in or logged out. it's that you assume only POST requests are made to this method. If I'm logged in and make a GET request, then your code will fail most likely.

prnvshrn replied:
I have added <b>if request.method is post </b> check in the delete method

prnvshrn commented:
I have made the following changes:
1. Changed file permission of <b>person.html</b> from 755 to 644
2. Added if condition to check whether request method is post before deleting investigator

jdamerow commented on browser/views.py:
```diff
@@ -624,3 +625,64 @@ def attendee_create(request, course_id):
+
+@staff_member_required
+def delete_investigator_record(request, person_id, research_id):
+    if request.method == 'POST':
```
and else?

prnvshrn replied:
User will be redirected to person page if a GET request is made for delete_investigator_record

prnvshrn commented:
Added the following changes:
1. User will be redirected to person page if a GET request is made for delete_investigator_record
